hashring/hashring.go

```diff
@@ -129,88 +135,117 @@ func (r *HashRing) AddServer(address string) bool {
 }
 
 // This function isn't thread-safe, only call it when the HashRing is locked.
-func (r *HashRing) addServerNoLock(address string) bool {
-	if _, ok := r.serverSet[address]; ok {
+func (r *HashRing) addServerNoLock(server membership.Member) bool {
+	if _, ok := r.serverSet[server.GetAddress()]; ok {
 		return false
 	}
 
-	r.addReplicasNoLock(address)
+	r.addReplicasNoLock(server)
 	return true
 }
 
 // This function isn't thread-safe, only call it when the HashRing is locked.
-func (r *HashRing) addReplicasNoLock(server string) {
-	r.serverSet[server] = struct{}{}
+func (r *HashRing) addReplicasNoLock(server membership.Member) {
+	r.serverSet[server.GetAddress()] = struct{}{}
 	for i := 0; i < r.replicaPoints; i++ {
-		address := fmt.Sprintf("%s%v", server, i)
-		r.tree.Insert(replicaPoint(r.hashfunc(address)), server)
+		r.tree.Insert(
+			r.replicaPointForServer(server, i),
+			server.GetAddress(),
+		)
 	}
 }
 
 // RemoveServer removes a server and its replicas from the HashRing.
-func (r *HashRing) RemoveServer(address string) bool {
+func (r *HashRing) RemoveServer(server membership.Member) bool {
 	r.Lock()
-	ok := r.removeServerNoLock(address)
+	ok := r.removeServerNoLock(server)
 	if ok {
 		r.computeChecksumNoLock()
 		r.EmitEvent(events.RingChangedEvent{
 			ServersAdded:   nil,
-			ServersRemoved: []string{address},
+			ServersRemoved: []string{server.GetAddress()},
 		})
 	}
 	r.Unlock()
 	return ok
 }
 
 // This function isn't thread-safe, only call it when the HashRing is locked.
-func (r *HashRing) removeServerNoLock(address string) bool {
-	if _, ok := r.serverSet[address]; !ok {
+func (r *HashRing) removeServerNoLock(server membership.Member) bool {
+	if _, ok := r.serverSet[server.GetAddress()]; !ok {
 		return false
 	}
 
-	r.removeReplicasNoLock(address)
+	r.removeReplicasNoLock(server)
 	return true
 }
 
 // This function isn't thread-safe, only call it when the HashRing is locked.
-func (r *HashRing) removeReplicasNoLock(server string) {
-	delete(r.serverSet, server)
+func (r *HashRing) removeReplicasNoLock(server membership.Member) {
+	delete(r.serverSet, server.GetAddress())
 	for i := 0; i < r.replicaPoints; i++ {
-		address := fmt.Sprintf("%s%v", server, i)
-		r.tree.Delete(replicaPoint(r.hashfunc(address)))
+		r.tree.Delete(
+			r.replicaPointForServer(server, i),
+		)
 	}
 }
 
+func (r *HashRing) ProcessMembershipChangesServers(changes []membership.MemberChange) {
+	r.Lock()
```

Comment:
Don't think we need it but it might be a performance improvement to first loop through the changes to check for relevant changes. If there are, lock, process them, compute checksum and unlock